### PR TITLE
Add an option to use axis-aligned geometry for hex/quad/lines in hype…

### PR DESCRIPTION
**Pull request by Davknapp**
…rcube_pad

Added a flag to use the axis-aligned geometrie for faster evaluation for hex/quads/lines

**_All these boxes must be checked by the reviewers before merging the pull request:_**

As a reviewer please read through all the code lines and make sure that the code is fully understood, bug free, well-documented and well-structured.

#### General
- [ ] The reviewer executed the new code features at least once and checked the results manually

- [ ] The code follows the [t8code coding guidelines](https://github.com/holke/t8code/wiki/Coding-Guideline)
- [ ] New source/header files are properly added to the Makefiles
- [ ] The code is well documented
- [ ] All function declarations, structs/classes and their members have a proper doxygen documentation
- [ ] All new algorithms and data structures are sufficiently optimal in terms of memory and runtime (If this should be merged, but there is still potential for optimization, create a new issue)

#### Tests
- [ ] The code is covered in an existing or new test case using Google Test

#### Github action

- [ ] The code compiles without warning in debugging and release mode, with and without MPI (this should be executed automatically in a github action)
- [ ] All tests pass (in various configurations, this should be executed automatically in a github action)

  If the Pull request introduces code that is not covered by the github action (for example coupling with a new library):
  - [ ] Should this use case be added to the github action?
  - [ ] If not, does the specific use case compile and all tests pass (check manually)

#### Scripts and Wiki

- [ ] If a new directory with source-files is added, it must be covered by the `script/find_all_source_files.scp` to check the indentation of these files.
- [ ] If this PR introduces a new feature, it must be covered in an example/tutorial and a Wiki article.

#### Licence

- [ ] The author added a BSD statement to `doc/` (or already has one)

### CODE REVIEW

File: src/t8_cmesh/t8_cmesh_examples.h [hunk 1]
```diff
@@ -132,7 +132,7 @@ t8_cmesh_new_hypercube (t8_eclass_t eclass, sc_MPI_Comm comm, int do_bcast, int
  */
 t8_cmesh_t
 t8_cmesh_new_hypercube_pad (const t8_eclass_t eclass, sc_MPI_Comm comm, const double *boundary, t8_locidx_t polygons_x,
-                            t8_locidx_t polygons_y, t8_locidx_t polygons_z);
+                            t8_locidx_t polygons_y, t8_locidx_t polygons_z, int use_axis_aligned);
```
Comment (sandro-elsweijer) [inline src/t8_cmesh/t8_cmesh_examples.h:135]: Please add the argument to the docstring

File: src/t8_cmesh/t8_cmesh_examples.c [hunk 2]
```diff
@@ -1215,9 +1216,16 @@ t8_cmesh_new_hypercube_pad (const t8_eclass_t eclass, sc_MPI_Comm comm, const do
   t8_cmesh_t cmesh;
   t8_cmesh_init (&cmesh);
 
-  /* We use standard linear geometry */
-  const t8_geometry_c *linear_geom = t8_geometry_linear_new (dim);
-  t8_cmesh_register_geometry (cmesh, linear_geom);
+  if (use_axis_aligned && (eclass == T8_ECLASS_HEX || eclass == T8_ECLASS_QUAD || eclass == T8_ECLASS_LINE)) {
+    /* We use axis aligned geometries */
+    const t8_geometry_c *axis_aligned_geom = t8_geometry_linear_axis_aligned_new (dim);
+    t8_cmesh_register_geometry (cmesh, axis_aligned_geom);
```
Comment (sandro-elsweijer) [inline src/t8_cmesh/t8_cmesh_examples.c:1222]: The axis aligned geometry uses only 2 vertices per tree. This has to be adapted as well

**Review by sandro-elsweijer — CHANGES_REQUESTED** [review]

File: src/t8_cmesh/t8_cmesh_examples.h (hunk 1 above)
Comment (Davknapp) [reply src/t8_cmesh/t8_cmesh_examples.h]: Done

File: src/t8_cmesh/t8_cmesh_examples.c (hunk 2 above)
Comment (Davknapp) [reply src/t8_cmesh/t8_cmesh_examples.c:1237]: Done

File: src/t8_cmesh/t8_cmesh_examples.c [hunk 3]
```diff
@@ -1078,36 +1088,42 @@ t8_cmesh_set_vertices_3D (t8_cmesh_t cmesh, const t8_eclass_t eclass, const doub
   for (t8_locidx_t hex_z_id = 0; hex_z_id < hexs_z; hex_z_id++) {
     for (t8_locidx_t hex_y_id = 0; hex_y_id < hexs_y; hex_y_id++) {
       for (t8_locidx_t hex_x_id = 0; hex_x_id < hexs_x; hex_x_id++) {
+        if (use_axis_aligned_geom && eclass == T8_ECLASS_HEX) {
```
Comment (sandro-elsweijer) [inline src/t8_cmesh/t8_cmesh_examples.c:1091]: Should we abort if `use_axis_aligned_geom && eclass != T8_ECLASS_HEX`?

File: src/t8_cmesh/t8_cmesh_examples.c [hunk 4]
```diff
@@ -940,21 +942,28 @@ t8_cmesh_set_vertices_2D (t8_cmesh_t cmesh, const t8_eclass_t eclass, const doub
   for (t8_locidx_t quad_y_id = 0; quad_y_id < quads_y; quad_y_id++) {
     for (t8_locidx_t quad_x_id = 0; quad_x_id < quads_x; quad_x_id++) {
 
-      t8_vec_axy (box, vertices, 1.0);                /* Vertex 0 */
-      t8_vec_axpyz (box, box_dir, vertices + 6, 1.0); /* Vertex 2 */
+      if (use_axis_aligned_geom && eclass == T8_ECLASS_QUAD) {
```
Comment (sandro-elsweijer) [inline src/t8_cmesh/t8_cmesh_examples.c:945]: Should we abort if `use_axis_aligned_geom && eclass != T8_ECLASS_QUAD`?

File: src/t8_cmesh/t8_cmesh_examples.c [hunk 5]
```diff
@@ -1215,9 +1231,16 @@ t8_cmesh_new_hypercube_pad (const t8_eclass_t eclass, sc_MPI_Comm comm, const do
   t8_cmesh_t cmesh;
   t8_cmesh_init (&cmesh);
 
-  /* We use standard linear geometry */
-  const t8_geometry_c *linear_geom = t8_geometry_linear_new (dim);
-  t8_cmesh_register_geometry (cmesh, linear_geom);
+  if (use_axis_aligned_geom && (eclass == T8_ECLASS_HEX || eclass == T8_ECLASS_QUAD || eclass == T8_ECLASS_LINE)) {
```
Comment (sandro-elsweijer) [inline src/t8_cmesh/t8_cmesh_examples.c:1234]: Should we abort if `use_axis_aligned_geom && !(eclass == T8_ECLASS_HEX || eclass == T8_ECLASS_QUAD || eclass == T8_ECLASS_LINE)`?

**Review by sandro-elsweijer — CHANGES_REQUESTED** [review]

File: src/t8_cmesh/t8_cmesh_examples.c (hunk 3 above)
Comment (Davknapp) [reply src/t8_cmesh/t8_cmesh_examples.c]: Resturctured the program a little and added this feature too.

File: src/t8_cmesh/t8_cmesh_examples.c (hunk 4 above)
Comment (Davknapp) [reply src/t8_cmesh/t8_cmesh_examples.c]: Resturctured the program a little and added this feature too.

File: src/t8_cmesh/t8_cmesh_examples.c (hunk 5 above)
Comment (Davknapp) [reply src/t8_cmesh/t8_cmesh_examples.c:1243]: Done

**Review by sandro-elsweijer — CHANGES_REQUESTED** [review]
Shouldn't the function also expect only 6 boundary coordinates (min x, y, z and max x, y, and z), if the axis-aligned geometry is used?

**Comment by jmark** [comment]
I just talked with David about this PR and the follow-up PR. I have some general issues about the way this performance feature is introduced in t8code.

**Comment by Davknapp** [comment]
> I just talked with @Davknapp about this PR and also the follow-up PR. I see some general problems with the way this performance feature is introduced into t8code.

If I remember correctly our discussion was mainly about the follow-up PR (#751). As we construct a cmesh here this PR is not meant to give an improved performance. It is only about the cmesh. The improved performance is introduced in #751.

**Comment by jmark** [comment]
> > I just talked with @Davknapp about this PR and also the follow-up PR. I see some general problems with the way this performance feature is introduced into t8code.
> 
> If I remember correctly our discussion was mainly about the follow-up PR (#751). As we construct a cmesh here this PR is not meant to give an improved performance. It is only about the cmesh. The improved performance is introduced in #751.

I agree. The actual performance improvement is implemented in the follow-up PR (#751). My point is that I want to avoid to introduce you proposed option to the function signature. I propose to move the point-search-operation into the geometry module and specialize this for the axis-aligned geometry. Than the performant implementation is used automatically when the cmesh geometry is set to the axis-aligned geometry.

**Comment by Davknapp** [comment]
> > > I just talked with @Davknapp about this PR and also the follow-up PR. I see some general problems with the way this performance feature is introduced into t8code.
> > 
> > 
> > If I remember correctly our discussion was mainly about the follow-up PR (#751). As we construct a cmesh here this PR is not meant to give an improved performance. It is only about the cmesh. The improved performance is introduced in #751.
> 
> I agree. The actual performance improvement is implemented in the follow-up PR (#751).
> 
> My point is that I want to avoid to introduce your proposed `use axis aligned geometry` flag to the function signature. I strongly suggest to move the point-search-operation into the geometry module and specialize this for the axis-aligned geometry. Than the performant implementation is used automatically when the cmesh geometry is set to the axis-aligned geometry.
> 
> This might mean we should start from scratch and create a new PRs, moving the point-search-operation to the geometry module first.

Yes, I completly agree. But this PR is used to construct a cmesh with an axis-algined geometry. It does not introduce any new functionality to the point-search-operation. Instead it adds the possibility to construct the cmesh_hypercube_pad with trees that are axis-aligned. 
I think your points are valid, but all related to #751? For the cmesh itself I think it is ok to introduce this flag (we actually need the possibility to construct an axis-aligned geometry in a project).

**Comment by sandro-elsweijer** [comment]
> > > I just talked with @Davknapp about this PR and also the follow-up PR. I see some general problems with the way this performance feature is introduced into t8code.
> > 
> > 
> > If I remember correctly our discussion was mainly about the follow-up PR (#751). As we construct a cmesh here this PR is not meant to give an improved performance. It is only about the cmesh. The improved performance is introduced in #751.
> 
> I agree. The actual performance improvement is implemented in the follow-up PR (#751).
> 
> My point is that I want to avoid to introduce your proposed `use axis aligned geometry` flag to the function signature. I strongly suggest to move the point-search-operation into the geometry module and specialize this for the axis-aligned geometry. Than the performant implementation is used automatically when the cmesh geometry is set to the axis-aligned geometry.
> 
> This might mean we should start from scratch and create a new PRs, moving the point-search-operation to the geometry module first.

This is a good thought, and the added `geom_is_axis_aligned` argument in the other PR is not user-friendly. But, as suggested in the other PR, there will be a way to ask the geometry itself, what type it is. Then the point inside check can perform that optimization automatically.

**Review by jmark — APPROVED** [review]

**Comment by jmark** [comment]
LGTM! @sandro-elsweijer How about you?

**Review by sandro-elsweijer — CHANGES_REQUESTED** [review]
My branch is now merged, and we can use the new geometry query

File: src/t8_cmesh/t8_cmesh_examples.c [hunk 6]
```diff
@@ -991,12 +993,23 @@ t8_cmesh_set_vertices_2D (t8_cmesh_t cmesh, const t8_eclass_t eclass, const doub
...
+      }
+#endif
 
       /* Map vertices of current quad on to respective trees inside. */
       if (eclass == T8_ECLASS_QUAD) {
         /* No mapping is required. */
         const t8_gloidx_t tree_id = quad_y_id * quads_x + quad_x_id;
-        t8_cmesh_set_tree_vertices (cmesh, tree_id, vertices, 4);
+        t8_cmesh_set_tree_vertices (cmesh, tree_id, vertices,
+                                    (use_axis_aligned_geom && eclass == T8_ECLASS_QUAD) ? 2 : 4);
```
[the remark this answers is not in the document]
Comment (Davknapp) [reply src/t8_cmesh/t8_cmesh_examples.c:1012]: What do you mean? There is no variable prod.

File: src/t8_cmesh/t8_cmesh_examples.c [hunk 7]
```diff
@@ -1257,9 +1282,9 @@ t8_cmesh_new_hypercube_pad (const t8_eclass_t eclass, sc_MPI_Comm comm, const do
   t8_cmesh_t cmesh;
   t8_cmesh_init (&cmesh);
 
-  /* We use standard linear geometry */
-  const t8_geometry_c *linear_geom = t8_geometry_linear_new (dim);
-  t8_cmesh_register_geometry (cmesh, linear_geom);
+  const int is_axis_aligned = t8_geom_is_linear_axis_aligned (geometry);
```
[the remark this answers is not in the document]
Comment (Davknapp) [reply src/t8_cmesh/t8_cmesh_examples.c:1285]: Why is that better than the original implementation?

[the remark this answers is not in the document]
Comment (Davknapp) [reply src/t8_cmesh/t8_cmesh_examples.c]: With the latest update of the geometry-class, t8_geometry_c now has an incomplete typedef.

File: src/t8_cmesh/t8_cmesh_examples.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (sandro-elsweijer) [reply src/t8_cmesh/t8_cmesh_examples.c]: prod = Production/release mode

File: src/t8_cmesh/t8_cmesh_examples.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (sandro-elsweijer) [reply src/t8_cmesh/t8_cmesh_examples.c]: You used a cpp function without including the cpp header file. There is also a c version of this function

**Review by sandro-elsweijer — APPROVED** [review]

**Pull request merged**